lib/blas_quda.cu

```diff
@@ -26,19 +26,19 @@ namespace quda {
       unsigned int minGridSize() const { return maxGridSize(); }
 
     public:
-      Blas(const coeff_t &a, const coeff_t &b, const coeff_t &c, ColorSpinorField &x,
-           ColorSpinorField &y, ColorSpinorField &z, ColorSpinorField &w, ColorSpinorField &v) :
+      template <typename Vx, typename Vy, typename Vz, typename Vw, typename Vv>
+      Blas(const coeff_t &a, const coeff_t &b, const coeff_t &c, Vx &x, Vy &y, Vz &z, Vw &w, Vv &v) :
         TunableGridStrideKernel2D(x, (x.IsComposite() ? x.CompositeDim() : 1) * x.SiteSubset()),
         f(a, b, c),
         nParity(vector_length_y),
         a(a),
         b(b),
         c(c),
-        x(x),
-        y(y),
-        z(z),
-        w(w),
-        v(v)
+        x(const_cast<ColorSpinorField&>(x)),
+        y(const_cast<ColorSpinorField&>(y)),
+        z(const_cast<ColorSpinorField&>(z)),
+        w(const_cast<ColorSpinorField&>(w)),
+        v(const_cast<ColorSpinorField&>(v))
```

Comment:
This is out of the scope of this PR, but we need to have these `const_cast`'s to have only one definition of `Blas` and `BlasArg` right? Is there a template trick that would get rid of them?

Reply:
It's definitely the plan to remove these `const_cast`s.  The immediate issue that I haven't solved yet is that we don't know the type of the input `x`, `y`, etc. fields since they are template arguments to the constructor so when declaring the types in the class there's no way to know if they should be const or not.

Reply:
Maybe some sort of variadic templated `Blas` would do? I will think more about this.

Reply:
👍